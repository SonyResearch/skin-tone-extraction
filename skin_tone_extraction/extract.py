"""Main entrypoint to extract skin tone from images."""

from skin_tone_extraction.image import MaskedImage
from skin_tone_extraction.methods.krishnapriya import KrishnapriyaMethod
from skin_tone_extraction.methods.merler import MerlerMethod
from skin_tone_extraction.methods.overall import OverallAverageMethod, OverallModeMethod
from skin_tone_extraction.methods.thong import ThongMethod

METHODS = {
    "thong": ThongMethod,
    "average": OverallAverageMethod,
    "mode": OverallModeMethod,
    "krishnapriya": KrishnapriyaMethod,
    "merler": MerlerMethod,
}
DEFAULT_METHOD = "average"


def extract_skin_tone(
    masked_image: MaskedImage,
    method: str = DEFAULT_METHOD,
    **kwargs,
) -> dict[str, float]:
    """Extract skin tone from a MaskedImage using the specified method.

    Args:
        masked_image (MaskedImage): The MaskedImage instance containing image and mask.
        method (str, optional): Extraction method to use. Options are
            'thong', 'average', 'mode', 'krishnapriya', 'merler'. Defaults to 'average'.
        **kwargs: Additional keyword arguments passed to the extraction method.

    Returns:
        The result of the skin tone extraction method.

    Raises:
        ValueError: If an unknown method is specified.
    """
    if method not in METHODS:
        raise ValueError(f"Unknown method: {method}")

    # Instantiate the extraction method with any provided kwargs
    extraction_method = METHODS[method](**kwargs)

    # Extract skin tone and return measurements
    result = extraction_method.extract(masked_image)
    return result.measurements


def extract_skin_tone_from_paths(
    img_path: str,
    mask_path: str,
    mask_value: int,
    method: str = DEFAULT_METHOD,
    **kwargs,
) -> dict[str, float]:
    """Extract skin tone from an image using the specified method.

    Args:
        img_path (str): Path to the image file.
        mask_path (str): Path to the mask file.
        mask_value (int): Value in the mask to select the region of interest.
        method (str, optional): Extraction method to use. Options are
            'thong', 'average', 'mode', 'krishnapriya', 'merler'. Defaults to 'average'.
        **kwargs: Additional keyword arguments passed to the extraction method.

    Returns:
        The result of the skin tone extraction method.

    Raises:
        ValueError: If an unknown method is specified.
    """
    # Create MaskedImage instance
    masked_image = MaskedImage(img_path, mask_path, mask_value)

    return extract_skin_tone(masked_image, method=method, **kwargs)
