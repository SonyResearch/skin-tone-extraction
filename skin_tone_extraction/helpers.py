"""Various smaller helper functions."""

import numpy as np
from skimage.color import rgb2lab

from skin_tone_extraction.image import MaskedImage
from skin_tone_extraction.metrics import get_corrected_ita, get_hue, get_ita
from skin_tone_extraction.visualizer import Visualizer


def mode_hist(x: np.ndarray, bins: int | str = "sturges") -> float:
    """Compute a histogram and return the mode (most frequent value).

    Args:
        x: Input array of values.
        bins: Method for determining histogram bins or number of bins
            (see numpy.histogram).
        plot: If True, plot the histogram.

    Returns:
        The mode value from the histogram.
    """
    hist, hist_bins = np.histogram(x, bins=bins)
    mode = hist_bins[hist.argmax()]

    return mode, hist_bins


def prepare_skin_tone_columns(skin):
    """Prepare comprehensive skin tone measurements from RGB skin pixel data.

    Args:
        skin: Array of RGB skin pixel values with shape (n_pixels, 3).

    Returns:
        Dictionary containing various skin tone measurements:
        - lum: L* values from CIELAB
        - lab_a: a* values from CIELAB
        - lab_b: b* values from CIELAB
        - hue: Hue angle values
        - ita: Individual Typology Angle values
        - cita: Corrected Individual Typology Angle values
        - red: Red channel values
        - green: Green channel values
        - blue: Blue channel values
    """
    if skin.size == 0:
        import warnings

        warnings.warn("Input 'skin' is empty. Returning empty arrays.")
        empty = np.empty(0)
        return {
            "lum": empty,
            "lab_a": empty,
            "lab_b": empty,
            "hue": empty,
            "ita": empty,
            "cita": empty,
            "red": empty,
            "green": empty,
            "blue": empty,
        }
    skin_lab = rgb2lab(skin)
    ita = get_ita(skin_lab[:, 0], skin_lab[:, 2])
    cita = get_corrected_ita(skin_lab[:, 0], skin_lab[:, 2])
    hue = get_hue(skin_lab[:, 1], skin_lab[:, 2])
    columns = {
        "lum": skin_lab[:, 0],
        "lab_a": skin_lab[:, 1],
        "lab_b": skin_lab[:, 2],
        "hue": hue,
        "ita": ita,
        "cita": cita,
        "red": skin[:, 0],
        "green": skin[:, 1],
        "blue": skin[:, 2],
    }
    return columns


def calculate_average(columns, std=True):
    """Calculate average values and optionally standard deviations for raw values.

    Args:
        columns: Dictionary of skin tone measurement arrays.
        std: Whether to also calculate standard deviations.

    Returns:
        Dictionary containing mean values and optionally standard deviations
        for each measurement type.
    """
    result = {}
    for key, values in columns.items():
        if len(values) == 0:
            result[key] = np.nan
            if std:
                result[f"{key}_std"] = np.nan
        else:
            result[key] = np.mean(values)
            if std:
                result[f"{key}_std"] = np.std(values)
    return result


def calculate_mode(columns, visualizer=None):
    """Calculate mode values for skin tone measurements.

    Args:
        columns: Dictionary of skin tone measurement arrays.
        visualizer: Optional Visualizer instance for displaying histograms.

    Returns:
        Dictionary containing mode values for each measurement type.
    """
    result = {}
    for key, values in columns.items():
        if len(values) == 0:
            result[key] = np.nan
        else:
            result[key], bins = mode_hist(values)

            if visualizer is not None:
                visualizer.visualize_mode(
                    metric_name=key, values=values, bins=bins, mode=result[key]
                )
    return result


def visualize_mask(
    img_path: str, mask_path: str, mask_value: int, bg_val: int = 128
) -> None:
    """Visualize only the masked pixels from the image for a given mask value."""
    img, binary_mask = MaskedImage(img_path, mask_path, mask_value).load()

    visualizer = Visualizer()
    visualizer.visualize_loaded_mask(
        img=img,
        binary_mask=binary_mask,
        label=f"Image Mask (mask_value: {mask_value})",
        bg_val=bg_val,
    )
