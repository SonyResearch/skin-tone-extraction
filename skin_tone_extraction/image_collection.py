"""Collection class for managing and defining multiple MaskedImage(s)."""

import glob
import os
import re
from typing import Iterator, List, Optional, Tuple, Union

from skin_tone_extraction.dataset import DatasetConfig

from .image import MaskedImage


class MaskedImageCollection:
    """A collection of MaskedImages."""

    allow_grayscale: Optional[bool] = None
    min_mask_coverage_warning: Optional[float] = None
    max_mask_coverage_warning: Optional[float] = None

    def __init__(
        self,
        images: Union[str, List[str], Tuple[str, ...]],
        masks: Union[str, List[str], Tuple[str, ...]],
        mask_value: int,
        image_ids: Optional[List[str]] = None,
    ):
        """Initialize the collection with image and mask sources.

        Args:
            images: Image source - directory path, pattern with {image_id},
                   list of file paths, or tuple of patterns/directories
            masks: Mask source - directory path, pattern with {image_id},
                  list of file paths, or tuple of patterns/directories
            mask_value: Integer value used to select skin region in mask
            image_ids: Optional list of image IDs to validate against

        Raises:
            ValueError: If validation fails or no matching files found
        """
        self.mask_value = mask_value

        # Parse and validate inputs
        self._image_paths, self._mask_paths, self._image_ids = self._parse_inputs(
            images, masks, image_ids
        )
        self._validate_inputs()

        # Cache for loaded MaskedImage instances
        self._loaded_images = {}

    @classmethod
    def from_dataset_config(
        cls, dataset_config: DatasetConfig
    ) -> "MaskedImageCollection":
        """Create a MaskedImageCollection from a dataset configuration.

        Args:
            dataset_config: Dataset configuration object with images, masks,
                mask_value, and optionally image_ids attributes

        Returns:
            MaskedImageCollection instance

        Raises:
            ValueError: If required fields are missing from dataset config
        """
        # Note: DatasetConfig might not always be correctly typed due to parallel
        # processing. We can only really rely on it as a dataclass.
        images = getattr(dataset_config, "images", None)
        masks = getattr(dataset_config, "masks", None)
        mask_value = getattr(dataset_config, "mask_value", None)
        image_ids = getattr(dataset_config, "image_ids", None)

        if images is None or masks is None or mask_value is None:
            raise ValueError(
                "Dataset config is missing required fields: "
                "images, masks, or mask_value"
            )

        # Create instance
        collection = cls(
            images=images,
            masks=masks,
            mask_value=mask_value,
            image_ids=image_ids,
        )

        # Apply optional configuration parameters
        allow_grayscale = getattr(dataset_config, "allow_grayscale", None)
        if allow_grayscale is not None:
            collection.allow_grayscale = allow_grayscale
        min_mask_coverage_warning = getattr(
            dataset_config, "min_mask_coverage_warning", None
        )
        if min_mask_coverage_warning is not None:
            collection.min_mask_coverage_warning = min_mask_coverage_warning
        max_mask_coverage_warning = getattr(
            dataset_config, "max_mask_coverage_warning", None
        )
        if max_mask_coverage_warning is not None:
            collection.max_mask_coverage_warning = max_mask_coverage_warning

        return collection

    def _parse_inputs(
        self,
        images: Union[str, List[str], Tuple[str, ...]],
        masks: Union[str, List[str], Tuple[str, ...]],
        image_ids: Optional[List[str]] = None,
    ) -> Tuple[List[str], List[str], Optional[List[str]]]:
        """Parse images and masks inputs into lists of file paths.

        Supports multiple input formats:
        - Strings: Scans directories for matching files, if they contain {image_id} this
            will be treated as a wildcard and extracted. Files will be searched based on
            images and image_ids will then be placed into the mask paths.
        - Tuples: Multiple strings to be parsed with their outputs concatenated.
        - Lists: Direct file path lists (must have matching lengths), will NOT be parsed

        Args:
            images: Image source - directory path, pattern with {image_id},
                   list of file paths, or tuple of patterns/directories
            masks: Mask source - directory path, pattern with {image_id},
                  list of file paths, or tuple of patterns/directories
            image_ids: Optional list of image IDs to validate against

        Returns:
            Tuple containing:
            - List of image file paths
            - List of mask file paths
            - List of image IDs (or None if not available)

        Raises:
            ValueError: If validation fails or no matching files found
        """

        def is_pattern(path):
            return isinstance(path, str) and "{image_id}" in path

        # If lists are provided, just return them
        if isinstance(images, list) and isinstance(masks, list):
            if len(images) != len(masks):
                raise ValueError("Length of images and masks lists must match.")
            # If image_ids is provided, check length
            if image_ids is not None and len(image_ids) != len(images):
                raise ValueError("Length of image_ids must match images/masks lists.")
            return images, masks, image_ids

        # If either images or masks is a tuple, make sure both are
        if isinstance(images, tuple) and not isinstance(masks, tuple):
            masks = (masks,)
        if not isinstance(images, tuple) and isinstance(masks, tuple):
            images = (images,)

        if isinstance(images, tuple) and isinstance(masks, tuple):
            # Tuples detected -> Recursively iterate over patterns and concat the output
            if len(images) != len(masks):
                raise ValueError("Length of images and masks tuples must match.")
            if image_ids is not None:
                raise ValueError("image_ids not supported with tuple inputs.")

            img_paths = []
            mask_paths = []
            image_ids = []

            for i, m in zip(images, masks):
                # Avoid infinite recursion
                if not isinstance(i, str) or not isinstance(m, str):
                    raise ValueError("Tuple elements must be strings.")
                res_images, res_masks, res_ids = self._parse_inputs(i, m)

                img_paths.extend(res_images)
                mask_paths.extend(res_masks)
                if res_ids:
                    image_ids.extend(res_ids)

            return img_paths, mask_paths, image_ids if image_ids else None
        else:
            if is_pattern(images) and is_pattern(masks):
                # Pattern mode
                img_pattern = images.replace("{image_id}", "*")
                img_files = sorted(glob.glob(img_pattern))
                if not img_files:
                    raise ValueError(f"No images found for pattern: {img_pattern}")

                # Extract image_id using regex
                pattern_regex = re.escape(images).replace("\\{image_id\\}", "(.+)")
                img_id_re = re.compile(f"^{pattern_regex}$")
                image_ids = []
                for f in img_files:
                    m = img_id_re.match(f)
                    if m:
                        image_ids.append(m.group(1))
                    else:
                        # Try matching just the basename
                        m = img_id_re.match(os.path.basename(f))
                        if m:
                            image_ids.append(m.group(1))
                        else:
                            image_ids.append(os.path.splitext(os.path.basename(f))[0])

                mask_files = [
                    masks.replace("{image_id}", image_id) for image_id in image_ids
                ]
                img_paths = img_files
                mask_paths = mask_files
            elif not is_pattern(images) and not is_pattern(masks):
                # Directory mode (read all files in the directories)
                img_files = sorted(
                    [f for f in os.listdir(images) if not f.startswith(".")]
                )
                mask_files = sorted(
                    [f for f in os.listdir(masks) if not f.startswith(".")]
                )
                # Match files by name (assume same filenames in both dirs)
                common_files = sorted(set(img_files) & set(mask_files))
                if not common_files:
                    raise ValueError("No matching files found in both directories.")
                img_paths = [os.path.join(images, f) for f in common_files]
                mask_paths = [os.path.join(masks, f) for f in common_files]
                image_ids = None
            else:
                raise ValueError(
                    "Both images and masks must be either directories or patterns "
                    "with {image_id}."
                )

        return img_paths, mask_paths, image_ids

    def _validate_inputs(self) -> None:
        """Validate that image and mask paths exist and match.

        Raises:
            ValueError: If validation fails
        """
        if len(self._image_paths) != len(self._mask_paths):
            # Check for matching lengths
            raise ValueError(
                "Number of images and masks must be the same. "
                f"Got {len(self._image_paths)} images and "
                f"{len(self._mask_paths)} masks."
            )

        if self._image_ids:
            # Check for matching lengths
            if len(self._image_ids) != len(self._image_paths):
                raise ValueError(
                    "Number of image IDs must match number of images. "
                    f"Got {len(self._image_ids)} IDs and "
                    f"{len(self._image_paths)} images."
                )

            # Check for duplicates
            num_duplicates = len(self._image_ids) - len(set(self._image_ids))
            if num_duplicates > 0:
                fraction_duplicates = num_duplicates / len(self._image_ids)
                raise ValueError(
                    "Duplicate image IDs found."
                    f"Share of duplicates: {fraction_duplicates:.3f}"
                )
        else:
            # If there are no image IDs, ensure all image paths are unique
            num_duplicates = len(self._image_paths) - len(set(self._image_paths))
            if num_duplicates > 0:
                fraction_duplicates = num_duplicates / len(self._image_paths)
                raise ValueError(
                    "Duplicate image paths found."
                    f"Share of duplicates: {fraction_duplicates:.3f}"
                )

    def __len__(self) -> int:
        """Return the number of images in the collection."""
        return len(self._image_paths)

    def __getitem__(self, index: int) -> MaskedImage:
        """Get a MaskedImage instance by index, creating it if necessary.

        Args:
            index: Index of the image to retrieve

        Returns:
            MaskedImage instance

        Raises:
            IndexError: If index is out of range
        """
        if index < 0 or index >= len(self._image_paths):
            raise IndexError("Index out of range")

        # Check if already loaded
        if index not in self._loaded_images:
            # Create MaskedImage instance
            image_id = self._image_ids[index] if self._image_ids else None
            self._loaded_images[index] = MaskedImage(
                img_path=self._image_paths[index],
                mask_path=self._mask_paths[index],
                mask_value=self.mask_value,
                id=image_id,
                collection=self,
            )

        return self._loaded_images[index]

    def __iter__(self) -> Iterator[MaskedImage]:
        """Iterate over all MaskedImage instances in the collection."""
        for i in range(len(self)):
            yield self[i]

    @property
    def image_paths(self) -> List[str]:
        """Get list of image file paths."""
        return self._image_paths.copy()

    @property
    def mask_paths(self) -> List[str]:
        """Get list of mask file paths."""
        return self._mask_paths.copy()

    @property
    def image_ids(self) -> Optional[List[str]]:
        """Get list of image IDs if available."""
        return self._image_ids.copy() if self._image_ids else None

    def get_id(self, index: int) -> Optional[str]:
        """Get image ID by index if available.

        Args:
            index: Index of the ID to retrieve

        Returns:
            Image ID string or None if not available

        Raises:
            IndexError: If index is out of range
        """
        if index < 0 or index >= len(self._image_paths):
            raise IndexError("Index out of range")

        return self._image_ids[index] if self._image_ids else None

    def clear_cache(self) -> None:
        """Clear all loaded MaskedImage instances to free memory."""
        self._loaded_images.clear()

    def slice(
        self, start: int = None, stop: int = None, step: int = None
    ) -> "MaskedImageCollection":
        """Create a new collection with a slice of the current images.

        Args:
            start: Start index for slicing
            stop: Stop index for slicing
            step: Step size for slicing

        Returns:
            New MaskedImageCollection with sliced data
        """
        sliced_image_paths = self._image_paths[start:stop:step]
        sliced_mask_paths = self._mask_paths[start:stop:step]
        sliced_image_ids = None
        if self._image_ids:
            sliced_image_ids = self._image_ids[start:stop:step]

        return MaskedImageCollection(
            images=sliced_image_paths,
            masks=sliced_mask_paths,
            mask_value=self.mask_value,
            image_ids=sliced_image_ids,
        )

    def __repr__(self) -> str:
        """String representation of the collection."""
        return (
            f"MaskedImageCollection(length={len(self)}, "
            f"mask_value={self.mask_value}, "
            f"cached={len(self._loaded_images)})"
        )
