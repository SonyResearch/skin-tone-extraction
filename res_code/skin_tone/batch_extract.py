"""Helper functions for parallel batch extraction."""

import os
from concurrent.futures import ThreadPoolExecutor, as_completed

import pandas as pd
from tqdm import tqdm

from .extract import extract_skin_tone
from .image import GrayscaleError
from .image_collection import MaskedImageCollection

RESULT_EMPTY = {}


def batch_extract(
    collection: MaskedImageCollection,
    max_workers: int = None,
    **kwargs,
) -> list[dict[str, float]]:
    """Batch extract skin values from multiple images in parallel with progress bar."""

    def _extract(masked_image):
        return extract_skin_tone(masked_image, **kwargs)

    results = [None] * len(collection)
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        futures = {
            executor.submit(_extract, collection[idx]): idx
            for idx in range(len(collection))
        }
        with tqdm(total=len(collection), desc="Extracting", smoothing=0) as pbar:
            for future in as_completed(futures):
                idx = futures[future]
                try:
                    results[idx] = future.result()
                except GrayscaleError as e:
                    masked_image = collection[idx]
                    print(
                        f"Warning: Skipping grayscale image "
                        f"'{masked_image.img_path}': {e}"
                    )
                    results[idx] = RESULT_EMPTY
                except Exception as e:
                    masked_image = collection[idx]
                    results[idx] = RESULT_EMPTY
                    print(
                        f"Exception for img_path='{masked_image.img_path}', "
                        f"mask_path='{masked_image.mask_path}'"
                    )
                    raise e
                pbar.update(1)
    return results


def batch_extract_df(
    collection: MaskedImageCollection,
    max_workers: int = None,
    expand_diagnostics: bool = True,
    **kwargs,
) -> pd.DataFrame:
    """Batch extract skin values and return as a DataFrame with metadata."""
    results = batch_extract(
        collection=collection,
        max_workers=max_workers,
        **kwargs,
    )
    results_df = pd.DataFrame(results)
    meta_df = pd.DataFrame(
        {
            "img_filename": [os.path.basename(p) for p in collection.image_paths],
            "mask_filename": [os.path.basename(p) for p in collection.mask_paths],
            "img_path": collection.image_paths,
            "mask_path": collection.mask_paths,
        }
    )
    df = pd.concat([meta_df, results_df], axis=1)

    if expand_diagnostics and "diagnostics" in df.columns:
        diag_df = df["diagnostics"].apply(pd.Series)
        if not diag_df.empty:
            diag_df = diag_df.add_prefix("diag_")
            df = pd.concat([df.drop(columns=["diagnostics"]), diag_df], axis=1)

    return df
