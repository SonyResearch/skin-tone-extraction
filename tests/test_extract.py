"""Tests for extraction functions."""

import glob
import os
import tempfile

import numpy as np
import pytest

from skin_tone_extraction.batch_extract import batch_extract_df
from skin_tone_extraction.extract import extract_skin_tone_from_paths
from skin_tone_extraction.helpers import (
    mode_hist,
    prepare_skin_tone_columns,
    visualize_mask,
)
from skin_tone_extraction.image_collection import MaskedImageCollection
from skin_tone_extraction.methods.overall import OverallAverageMethod
from skin_tone_extraction.methods.thong import ThongMethod
from skin_tone_extraction.metrics import get_hue
from skin_tone_extraction.visualizer import Visualizer

ASSETS_DIR = os.path.join(os.path.dirname(__file__), "./assets")
IMG_PATH = os.path.abspath(os.path.join(ASSETS_DIR, "00000.png"))
MASK_PATH = os.path.abspath(os.path.join(ASSETS_DIR, "00000_mask.png"))


def test_extract_image_basic():
    result = extract_skin_tone_from_paths(
        IMG_PATH, MASK_PATH, mask_value=1, method="thong"
    )

    # Hard-coded values based on example in https://github.com/SonyResearch/apparent_skincolor
    expected_values = {
        "lum": 60.32474192162227,
        "hue": 27.9874036146538,
        "lum_std": 7.537910024242527,
        "hue_std": 68.11535999823427,
        "red": 0.5670517343182768,
        "green": 0.5587551551471699,
        "blue": 0.4942237944973208,
        "red_std": 0.09923820062668536,
        "green_std": 0.059575927609366405,
        "blue_std": 0.055943684718035125,
    }

    # Check if all expected keys are present
    for key in expected_values:
        assert key in result, f"Missing key: {key}"
        assert isinstance(result[key], float), f"{key} is not a float"

    # Compare values with expected values
    for key, expected_value in expected_values.items():
        assert abs(result[key] - expected_value) < 1e-5, (
            f"Value for {key} does not match: {result[key]} != {expected_value}"
        )


def test_extract_snapshot_thong(snapshot):
    result = extract_skin_tone_from_paths(
        IMG_PATH, MASK_PATH, mask_value=1, method="thong"
    )
    assert result == snapshot


def test_extract_snapshot_average(snapshot):
    result = extract_skin_tone_from_paths(
        IMG_PATH, MASK_PATH, mask_value=1, method="average"
    )
    assert result == snapshot


def test_extract_snapshot_mode(snapshot):
    result = extract_skin_tone_from_paths(
        IMG_PATH, MASK_PATH, mask_value=1, method="mode"
    )
    assert result == snapshot


def test_extract_snapshot_merler(snapshot):
    result = extract_skin_tone_from_paths(
        IMG_PATH, MASK_PATH, mask_value=1, method="merler"
    )
    assert result == snapshot


def test_extract_snapshot_krishnapriya(snapshot):
    result = extract_skin_tone_from_paths(
        IMG_PATH, MASK_PATH, mask_value=1, method="krishnapriya"
    )
    assert result == snapshot


def test_extract_snapshot_merler_nonsmooth_avg(snapshot):
    result = extract_skin_tone_from_paths(
        IMG_PATH,
        MASK_PATH,
        mask_value=1,
        method="merler",
        smooth=False,
        aggregate_regions="average",
    )
    assert result == snapshot


def test_get_hue():
    a = np.array([1.0, 0.0, -1.0])
    b = np.array([0.0, 1.0, 0.0])
    hue = get_hue(a, b)
    assert hue.shape == a.shape
    assert np.all(np.isfinite(hue))


def test_mode_hist():
    x = np.array([1, 2, 2, 3, 3, 3, 4])
    mode, _ = mode_hist(x, bins=4)
    assert isinstance(mode, float)


def test_clustering():
    x = np.random.rand(10, 2)
    method = ThongMethod()
    labels, model = method.clustering(x, n_clusters=2)
    assert labels.shape[0] == 10
    assert hasattr(model, "predict")


def test_get_scalar_values():
    # Use synthetic LAB data and labels
    lab = np.random.rand(10, 3) * 100
    rgb = np.random.rand(10, 3) * 100
    labels = np.array([0, 1] * 5)
    method = ThongMethod()
    res = method.get_scalar_values(lab, labels, rgb, topk=1, skip_topk=0)
    assert "lum" in res and "hue" in res


def test_visualize_mask():
    # Should not raise
    visualize_mask(IMG_PATH, MASK_PATH, mask_value=1)


def test_visualize_clusters_on_image():
    img = np.ones((5, 5, 3), dtype=np.float32)
    mask = np.ones((5, 5), dtype=bool)
    labels = np.zeros(mask.sum(), dtype=int)
    viz = Visualizer()
    viz.visualize_clusters_on_image(labels, img, mask)


def test_show():
    img = np.ones((5, 5, 3), dtype=np.float32)
    viz = Visualizer()
    viz.visualize_image(img, title="Test Show")


def test_calculate_average_skin_tone_empty():
    empty = np.array([])
    method = OverallAverageMethod()
    with pytest.warns(Warning):
        result = method.calculate_average_skin_tone(empty)
    # Should return means and stds as nan or handle gracefully
    assert all(np.isnan(v) for v in result.values())


def test_prepare_skin_tone_columns_empty():
    empty = np.array([])
    with pytest.warns(Warning):
        columns = prepare_skin_tone_columns(empty)
    # Should return dict with empty arrays for each color channel
    assert isinstance(columns, dict)
    for v in columns.values():
        assert isinstance(v, np.ndarray)
        assert v.size == 0


def test_batch_extract_df_basic():
    # Create MaskedImageCollection
    collection = MaskedImageCollection(
        images=[IMG_PATH], masks=[MASK_PATH], mask_value=1
    )
    df = batch_extract_df(
        collection=collection, method="merler", expand_diagnostics=True
    )
    # Check DataFrame shape
    assert df.shape[0] == 1
    # Check metadata columns
    for col in ["img_filename", "mask_filename", "img_path", "mask_path"]:
        assert col in df.columns
    # Check skin tone columns
    for col in [
        "lum",
        "hue",
        "lum_std",
        "hue_std",
        "red",
        "green",
        "blue",
        "red_std",
        "green_std",
        "blue_std",
    ]:
        assert col in df.columns
    # Check values are floats
    for col in [
        "lum",
        "hue",
        "lum_std",
        "hue_std",
        "red",
        "green",
        "blue",
        "red_std",
        "green_std",
        "blue_std",
    ]:
        assert isinstance(df.iloc[0][col], float)
    # Check whether columns exist with the diag_ prefix
    assert "diagnostics" not in df.columns, "Diagnostics column should not be present"
    diag_cols = [col for col in df.columns if col.startswith("diag_")]
    assert len(diag_cols) > 0, "No expanded diagnostic columns found"


def test_debug_image_saving():
    """Test that debug images are saved when debug=True and debug_image_dir is set."""
    # Create a temporary directory for debug images
    with tempfile.TemporaryDirectory() as temp_dir:
        # Extract skin tone with debug enabled
        result = extract_skin_tone_from_paths(
            IMG_PATH,
            MASK_PATH,
            mask_value=1,
            method="thong",
            debug=True,
            debug_image_dir=temp_dir,
        )

        # Check that result contains expected measurements
        assert isinstance(result, dict)
        assert "lum" in result
        assert "hue" in result

        # Check that debug images were saved
        saved_files = glob.glob(os.path.join(temp_dir, "*.png"))
        assert len(saved_files) > 0, "No debug images were saved"

        # Check for specific expected debug images based on ThongMethod
        # visualization calls. The ThongMethod creates these visualizations
        # when debug=True:
        # - "Original" (via visualize_image)
        # - "Smoothed" (via visualize_image)
        # - "Smoothed (Masked)" (via visualize_loaded_mask)
        # - "Lum" (via visualize_1dim_mask)
        # - "Hue" (via visualize_1dim_mask)
        # - clusters visualization (via visualize_clusters_on_image)

        # Check for at least some expected debug image files
        expected_patterns = [
            "*original*.png",
            "*smoothed*.png",
            "*lum*.png",
            "*hue*.png",
        ]

        found_patterns = []
        for pattern in expected_patterns:
            matches = glob.glob(os.path.join(temp_dir, pattern))
            if matches:
                found_patterns.append(pattern)

        # Ensure at least some debug images were created
        saved_file_names = [os.path.basename(f) for f in saved_files]
        assert len(found_patterns) > 0, (
            f"Expected debug image patterns not found. Saved files: {saved_file_names}"
        )
