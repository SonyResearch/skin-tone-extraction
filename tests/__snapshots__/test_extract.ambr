# serializer version: 1
# name: test_extract_snapshot_average
  dict({
    'blue': 0.39258658926976125,
    'blue_std': 0.14058162590331652,
    'cita': 5.108862232771542,
    'cita_std': 45.783523713671514,
    'green': 0.4596805985992942,
    'green_std': 0.11595653465483254,
    'hue': 20.592885891606027,
    'hue_std': 71.28570685584526,
    'ita': 8.11241643220028,
    'ita_std': 45.34776970450102,
    'lab_a': 0.22817850410144763,
    'lab_a_std': 3.9823243856886306,
    'lab_b': 10.769957829575368,
    'lab_b_std': 7.401241639723248,
    'lum': 49.368036017412365,
    'lum_std': 11.93167927985044,
    'red': 0.48977307826269806,
    'red_std': 0.12359386222698164,
  })
# ---
# name: test_extract_snapshot_krishnapriya
  dict({
    'blue': 0.3081298508955549,
    'cita': -21.477870616519493,
    'diagnostics': dict({
      'filter_frac': 0.2584890285015805,
      'n_pixels': 78667,
    }),
    'green': 0.3958503665363887,
    'hue': 76.39050071194713,
    'ita': -21.477870616519493,
    'lab_a': 3.729421967587687,
    'lab_b': 15.404407706519708,
    'lum': 43.93891421913617,
    'red': 0.46772379080154386,
  })
# ---
# name: test_extract_snapshot_merler
  dict({
    'blue': 0.47573669317021755,
    'blue_std': 0.10940410281839831,
    'cita': 19.655670783101566,
    'cita_std': 51.639185242591736,
    'diagnostics': dict({
      'region_bounds': list([
        tuple(
          tuple(
            164,
            372,
          ),
          tuple(
            146,
            512,
          ),
        ),
        tuple(
          tuple(
            164,
            372,
          ),
          tuple(
            512,
            745,
          ),
        ),
        tuple(
          tuple(
            372,
            581,
          ),
          tuple(
            146,
            512,
          ),
        ),
        tuple(
          tuple(
            372,
            581,
          ),
          tuple(
            512,
            745,
          ),
        ),
      ]),
      'region_sizes': list([
        52830,
        40210,
        62584,
        33333,
      ]),
      'total_pixels': 188957,
    }),
    'green': 0.5268705654760313,
    'green_std': 0.09545656742171126,
    'hue': 65.68432264841385,
    'hue_std': 4.607300436177647,
    'ita': 19.675269765940786,
    'ita_std': 48.902815804164305,
    'lab_a': 1.3400396786926227,
    'lab_a_std': 1.860224673823882,
    'lab_b': 7.2953866506986795,
    'lab_b_std': 2.6534797188254142,
    'lum': 56.06572772850886,
    'lum_std': 9.058698662369892,
    'red': 0.5314325132320602,
    'red_std': 0.11508025579582294,
  })
# ---
# name: test_extract_snapshot_merler_nonsmooth_avg
  dict({
    'blue': 112.32534386130177,
    'blue_std': 17.951208249170406,
    'cita': 15.264435944696935,
    'cita_std': 32.7882293024946,
    'diagnostics': dict({
      'region_bounds': list([
        tuple(
          tuple(
            164,
            372,
          ),
          tuple(
            146,
            512,
          ),
        ),
        tuple(
          tuple(
            164,
            372,
          ),
          tuple(
            512,
            745,
          ),
        ),
        tuple(
          tuple(
            372,
            581,
          ),
          tuple(
            146,
            512,
          ),
        ),
        tuple(
          tuple(
            372,
            581,
          ),
          tuple(
            512,
            745,
          ),
        ),
      ]),
      'region_sizes': list([
        52830,
        40210,
        62584,
        33333,
      ]),
      'total_pixels': 188957,
    }),
    'green': 123.91845234630101,
    'green_std': 17.29010744313991,
    'hue': 37.33040807610395,
    'hue_std': 14.96584996249815,
    'ita': 19.307735574665415,
    'ita_std': 23.276200495122033,
    'lab_a': 0.9817164819434936,
    'lab_a_std': 1.635425648805309,
    'lab_b': 7.551253317427368,
    'lab_b_std': 0.7361880092274128,
    'lum': 52.126877480066284,
    'lum_std': 7.245316492369242,
    'red': 131.20375005953736,
    'red_std': 21.76452244882815,
  })
# ---
# name: test_extract_snapshot_mode
  dict({
    'blue': 0.2,
    'cita': -36.00009254757575,
    'green': 0.36,
    'hue': 80.99956121564693,
    'ita': -36.00009254757575,
    'lab_a': 1.6756546053227623,
    'lab_b': 6.158236609761854,
    'lum': 40.946752869796555,
    'red': 0.6,
  })
# ---
# name: test_extract_snapshot_thong
  dict({
    'blue': 0.49422379449732085,
    'blue_std': 0.05594368471803509,
    'cita': 54.252207458317336,
    'cita_std': 16.834894838739412,
    'green': 0.5587551551471699,
    'green_std': 0.0595759276093664,
    'hue': 27.9874036146538,
    'hue_std': 68.11535999823427,
    'ita': 56.59046833045702,
    'ita_std': 12.699870095061462,
    'lab_a': 0.6421512575989494,
    'lab_a_std': 1.0049845697630662,
    'lab_b': 4.800789388688733,
    'lab_b_std': 2.020569079079817,
    'lum': 60.32474192162227,
    'lum_std': 7.537910024242527,
    'red': 0.567051734318277,
    'red_std': 0.0992382006266855,
  })
# ---
